File: context/lattice/test/sample.py
"""
Sample Python module for testing the parser.
Contains various Python constructs for comprehensive parsing.
"""

import os
import sys
from typing import Optional, List, Dict, Any
from dataclasses import dataclass
from pathlib import Path
from .services import UserService, AuthService
from ..utils import helpers as utils
import requests


@dataclass
class User:
    """User data model."""
    id: int
    name: str
    email: str
    role: str = "user"


class UserController:
    """
    Controller for user-related operations.
    Handles HTTP requests for user management.
    """
    
    def __init__(self, user_service: UserService, auth_service: AuthService):
        """Initialize the controller with required services."""
        self.user_service = user_service
        self.auth_service = auth_service
        self._cache: Dict[str, Any] = {}
    
    async def get_user(self, request, response) -> Optional[Dict]:
        """
        Get user by ID.
        
        Args:
            request: The HTTP request object
            response: The HTTP response object
            
        Returns:
            User data dictionary or None if not found
        """
        user_id = request.params.get('id')
        user = await self.user_service.find_by_id(user_id)
        
        if not user:
            response.status(404).json({'error': 'User not found'})
            return None
        
        return response.json({'success': True, 'data': user})
    
    async def create_user(self, request, response) -> Dict:
        """Create a new user."""
        data = request.body
        name = data.get('name')
        email = data.get('email')
        role = data.get('role', 'user')
        
        if not self._validate_email(email):
            response.status(400).json({'error': 'Invalid email'})
            return None
        
        user = await self.user_service.create({
            'name': name,
            'email': email,
            'role': role
        })
        
        return response.status(201).json({'success': True, 'data': user})
    
    def _validate_email(self, email: str) -> bool:
        """Validate email format."""
        import re
        pattern = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
        return bool(re.match(pattern, email))
    
    @staticmethod
    def format_user_response(user: User) -> Dict:
        """Format user object for API response."""
        return {
            'id': user.id,
            'name': user.name,
            'email': user.email,
            'role': user.role
        }
    
    @property
    def cache_size(self) -> int:
        """Get current cache size."""
        return len(self._cache)


class DataUtils:
    """Utility functions for data processing."""
    
    @staticmethod
    def array_to_map(arr: List[Dict], key: str) -> Dict[str, Dict]:
        """Transform array to map by key."""
        result = {}
        for item in arr:
            result[str(item[key])] = item
        return result
    
    @staticmethod
    def deep_clone(obj: Any) -> Any:
        """Deep clone an object."""
        import json
        return json.loads(json.dumps(obj))


async def initialize_database(connection_string: str) -> bool:
    """
    Initialize database connection.
    
    Args:
        connection_string: Database connection URL
        
    Returns:
        True if connection successful, False otherwise
    """
    import asyncio
    try:
        print('Connecting to database...')
        await asyncio.sleep(1)
        print('Database connected')
        return True
    except Exception as error:
        print(f'Database connection failed: {error}')
        return False


async def fetch_external_data(endpoint: str) -> Any:
    """Fetch data from external API."""
    response = requests.get(endpoint)
    return response.json()


def calculate_total(items: List[int]) -> int:
    """Calculate sum of items."""
    return sum(items)


def format_currency(amount: float, currency: str = 'USD') -> str:
    """Format number as currency string."""
    symbols = {'USD': '$', 'EUR': '€', 'GBP': '£'}
    symbol = symbols.get(currency, currency)
    return f'{symbol}{amount:,.2f}'


@property
def config_path() -> Path:
    """Get configuration file path."""
    return Path.home() / '.config' / 'app'


if __name__ == '__main__':
    import asyncio
    
    async def main():
        success = await initialize_database('postgresql://localhost/mydb')
        if success:
            print('Application started')
    
    asyncio.run(main())

File: context/lattice/test/sample.ts
/**
 * Sample TypeScript Test File
 * Contains various code patterns for testing the parser
 */

import express, { Request, Response } from 'express';
import { UserService, AuthService } from './services';
import * as utils from '../utils';
import axios from 'axios';

const fs = require('fs');
const { promisify } = require('util');

// Type definitions
interface User {
  id: string;
  name: string;
  email: string;
  role: UserRole;
}

type UserRole = 'admin' | 'user' | 'guest';

type ApiResponse<T> = {
  success: boolean;
  data: T;
  error?: string;
};

/**
 * @class UserController
 * Handles all user-related HTTP endpoints
 */
class UserController {
  private userService: UserService;
  private authService: AuthService;

  constructor(userService: UserService, authService: AuthService) {
    this.userService = userService;
    this.authService = authService;
  }

  /**
   * Get user by ID
   * @param {Request} req - Express request
   * @param {Response} res - Express response
   * @returns {Promise<void>}
   */
  async getUser(req: Request, res: Response): Promise<void> {
    const { id } = req.params;
    const user = await this.userService.findById(id);
    
    if (!user) {
      res.status(404).json({ error: 'User not found' });
      return;
    }
    
    res.json({ success: true, data: user });
  }

  /**
   * Create new user
   * @param {Request} req - Express request with user data
   * @param {Response} res - Express response
   */
  async createUser(req: Request, res: Response) {
    const { name, email, role } = req.body;
    
    if (!this.validateEmail(email)) {
      res.status(400).json({ error: 'Invalid email' });
      return;
    }

    const user = await this.userService.create({ name, email, role });
    res.status(201).json({ success: true, data: user });
  }

  private validateEmail(email: string): boolean {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
  }

  _internalMethod() {
    console.log('Internal processing');
  }
}

/**
 * Utility functions for data processing
 */
const DataUtils = {
  /**
   * Transform array to map
   */
  arrayToMap<T>(arr: T[], key: keyof T): Map<string, T> {
    const map = new Map();
    for (const item of arr) {
      map.set(String(item[key]), item);
    }
    return map;
  },

  /**
   * Deep clone object
   */
  deepClone<T>(obj: T): T {
    return JSON.parse(JSON.stringify(obj));
  },
};

/**
 * Initialize database connection
 * @param {string} connectionString - Database connection URL
 * @returns {Promise<boolean>} Connection success status
 */
async function initializeDatabase(connectionString: string): Promise<boolean> {
  try {
    console.log('Connecting to database...');
    await new Promise(resolve => setTimeout(resolve, 1000));
    console.log('Database connected');
    return true;
  } catch (error) {
    console.error('Database connection failed:', error);
    return false;
  }
}

/**
 * Fetch data from external API
 */
const fetchExternalData = async (endpoint: string) => {
  const response = await axios.get(endpoint);
  return response.data;
};

// Arrow function with implicit return
const calculateTotal = (items: number[]) => items.reduce((a, b) => a + b, 0);

// Regular function
function formatCurrency(amount: number, currency: string = 'USD'): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
  }).format(amount);
}

// Self-invoking setup
(async () => {
  const connected = await initializeDatabase('mongodb://localhost:27017');
  if (connected) {
    console.log('Application ready');
  }
})();

// Event handlers
process.on('SIGTERM', () => {
  console.log('Shutting down gracefully');
  process.exit(0);
});

// Exports
export { UserController, DataUtils, initializeDatabase };
export default UserController;
